code/modules/cm_marines/equipment/sentries.dm

```diff
@@ -918,29 +1050,57 @@
 
 	var/list/turf/path = list()
 	var/turf/T
-	var/mob/M
+	var/mob/living/M
 
 	for(M in oview(range, src))
-		if(!isliving(M) || M.stat || isrobot(M)) continue //No unconscious/deads, or non living.
-
+		if(M.stat == DEAD || isrobot(M)) //No dead or robots.
+			continue
+		if(!safety_off && !isXeno(M)) //When safeties are on, Xenos only.
+			continue
 		/*
 		I really, really need to replace this with some that isn't insane. You shouldn't have to fish for access like this.
 		This should be enough shortcircuiting, but it is possible for the code to go all over the possibilities and generally
 		slow down. It'll serve for now.
 		*/
 		var/mob/living/carbon/human/H = M
-		if(istype(H) && H.get_target_lock(iff_signal)) continue
+		if(istype(H) && H.get_target_lock(iff_signal))
+			continue
+
+
 
 		var/angle = get_dir(src, M)
-		if(angle & dir)
-			path = getline2(src, M)
+		if(angle & dir || radial_mode)
+			path = getline2(src, M, TRUE)
+			if(alerts_on) //They're within our field of detection and thus can trigger the alarm
+				if(world.time > (last_alert + SENTRY_ALERT_DELAY) || !(M in alert_list)) //if we're not on cooldown or the target isn't in the list, sound the alarm
+					playsound(loc, 'sound/machines/warning-buzzer.ogg', 50, FALSE)
+					sentry_alert(SENTRY_ALERT_HOSTILE, M)
+					alert_list.Add(M)
+					last_alert = world.time
 		else
 			continue
 
 		if(path.len)
+			var/blocked = FALSE
+			var/current_turf = get_turf(src)
 			for(T in path)
-				if(T.opacity) continue
-			targets += M
+				if(T == current_turf) //Don't deal with the current turf
```

Comment:
`path -= current_turf` before the `for` loop should offer better performance. Does the check only once.

Reply:
Or better even, after `path = getline2(src, M, TRUE)`

Reply:
I added this because an earlier change I made to getline2 didn't seem to work, but upon further testing it apparently does; this is handled.